From a8d62012e2b47e8f62fafe17ab1042b0415a06ea Mon Sep 17 00:00:00 2001
From: Rajalakshmi Srinivasaraghavan <raji@linux.vnet.ibm.com>
Date: Wed, 30 Jul 2014 05:20:37 -0500
Subject: [PATCH] PowerPC: multiarch memchr for PowerPC64

commit 870f867648fc62e230dad80cd34bd599f08e2193
Author: Adhemerval Zanella <azanella@linux.vnet.ibm.com>
Date:   Fri Dec 13 14:35:28 2013 -0500

Added string/memchr.c and
sysdeps/powerpc/powerpc32/power4/multiarch/memchr-ppc32.c aart from
original commit
---
 string/memchr.c                                    |  6 +++-
 .../powerpc32/power4/multiarch/memchr-ppc32.c      | 34 ++++++++++++++++++
 sysdeps/powerpc/powerpc64/multiarch/Makefile       |  2 +-
 .../powerpc/powerpc64/multiarch/ifunc-impl-list.c  |  8 +++++
 .../powerpc/powerpc64/multiarch/memchr-power7.S    | 40 ++++++++++++++++++++++
 sysdeps/powerpc/powerpc64/multiarch/memchr-ppc64.c | 19 ++++++++++
 sysdeps/powerpc/powerpc64/multiarch/memchr.c       | 38 ++++++++++++++++++++
 7 files changed, 145 insertions(+), 2 deletions(-)
 create mode 100644 sysdeps/powerpc/powerpc32/power4/multiarch/memchr-ppc32.c
 create mode 100644 sysdeps/powerpc/powerpc64/multiarch/memchr-power7.S
 create mode 100644 sysdeps/powerpc/powerpc64/multiarch/memchr-ppc64.c
 create mode 100644 sysdeps/powerpc/powerpc64/multiarch/memchr.c

diff --git glibc-2.17-c758a686/string/memchr.c glibc-2.17-c758a686/string/memchr.c
index 22637cf..822c0a4 100644
--- glibc-2.17-c758a686/string/memchr.c
+++ glibc-2.17-c758a686/string/memchr.c
@@ -56,9 +56,13 @@
 #undef memchr
 #undef __memchr
 
+#ifndef MEMCHR
+# define MEMCHR __memchr
+#endif
+
 /* Search no more than N bytes of S for C.  */
 __ptr_t
-__memchr (s, c_in, n)
+MEMCHR (s, c_in, n)
      const __ptr_t s;
      int c_in;
      size_t n;
diff --git glibc-2.17-c758a686/sysdeps/powerpc/powerpc32/power4/multiarch/memchr-ppc32.c glibc-2.17-c758a686/sysdeps/powerpc/powerpc32/power4/multiarch/memchr-ppc32.c
new file mode 100644
index 0000000..43c5652
--- /dev/null
+++ glibc-2.17-c758a686/sysdeps/powerpc/powerpc32/power4/multiarch/memchr-ppc32.c
@@ -0,0 +1,34 @@
+/* PowerPC32 default implementation of memchr.
+   Copyright (C) 2013-2014 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <string.h>
+
+#define MEMCHR  __memchr_ppc
+
+#undef weak_alias
+#define weak_alias(a, b)
+
+#ifdef SHARED
+# undef libc_hidden_builtin_def
+# define libc_hidden_builtin_def(name)  \
+  __hidden_ver1 (__memchr_ppc, __GI_memchr, __memchr_ppc);
+#endif
+
+extern __typeof (memchr) __memchr_ppc attribute_hidden;
+
+#include <string/memchr.c>
diff --git glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/Makefile glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/Makefile
index 22560e8..b2815e7 100644
--- glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/Makefile
+++ glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/Makefile
@@ -3,5 +3,5 @@ sysdep_routines += memcpy-power7 memcpy-a2 memcpy-power6 memcpy-cell \
                   memcpy-power4 memcpy-ppc64 memcmp-power7 memcmp-power4 \
                   memcmp-ppc64 memset-power7 memset-power6 memset-power4 \
                   memset-ppc64 bzero-power4 bzero-power6 bzero-power7 \
-                  mempcpy-power7 mempcpy-ppc64
+                  mempcpy-power7 mempcpy-ppc64 memchr-power7 memchr-ppc64
 endif
diff --git glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/ifunc-impl-list.c glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/ifunc-impl-list.c
index c72c229..663e294 100644
--- glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/ifunc-impl-list.c
+++ glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/ifunc-impl-list.c
@@ -98,5 +98,13 @@ __libc_ifunc_impl_list (const char *name, struct libc_ifunc_impl *array,
              IFUNC_IMPL_ADD (array, i, mempcpy, 1,
                              __mempcpy_ppc))
 
+  /* Support sysdeps/powerpc/powerpc64/multiarch/memchr.c.  */
+  IFUNC_IMPL (i, name, memchr,
+             IFUNC_IMPL_ADD (array, i, memchr,
+                             hwcap & PPC_FEATURE_HAS_VSX,
+                             __memchr_power7)
+             IFUNC_IMPL_ADD (array, i, memchr, 1,
+                             __memchr_ppc))
+
   return i;
 }
diff --git glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr-power7.S glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr-power7.S
new file mode 100644
index 0000000..b1b4ec7
--- /dev/null
+++ glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr-power7.S
@@ -0,0 +1,40 @@
+/* Optimized memchr implementation for PowerPC64/POWER7.
+   Copyright (C) 2013-2014 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <sysdep.h>
+
+#undef ENTRY
+#define ENTRY(name)						\
+  .section ".text";						\
+  ENTRY_2(__memchr_power7)					\
+  .align ALIGNARG(2);						\
+  BODY_LABEL(__memchr_power7):					\
+  cfi_startproc;
+
+#undef END
+#define END(name)						\
+  cfi_endproc;							\
+  TRACEBACK(__memchr_power7)					\
+  END_2(__memchr_power7)
+
+#undef libc_hidden_builtin_def
+#define libc_hidden_builtin_def(name)
+#undef weak_alias
+#define weak_alias(name,alias)
+
+#include <sysdeps/powerpc/powerpc64/power7/memchr.S>
diff --git glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr-ppc64.c glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr-ppc64.c
new file mode 100644
index 0000000..9e2a711
--- /dev/null
+++ glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr-ppc64.c
@@ -0,0 +1,19 @@
+/* PowerPC64 default implementation of memchr.
+   Copyright (C) 2013-2014 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#include <sysdeps/powerpc/powerpc32/power4/multiarch/memchr-ppc32.c>
diff --git glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr.c glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr.c
new file mode 100644
index 0000000..ca0f714
--- /dev/null
+++ glibc-2.17-c758a686/sysdeps/powerpc/powerpc64/multiarch/memchr.c
@@ -0,0 +1,38 @@
+/* Multiple versions of memchr.
+   Copyright (C) 2013-2014 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <http://www.gnu.org/licenses/>.  */
+
+#ifndef NOT_IN_libc
+# include <string.h>
+# include <shlib-compat.h>
+# include "init-arch.h"
+
+extern __typeof (__memchr) __memchr_ppc attribute_hidden;
+extern __typeof (__memchr) __memchr_power7 attribute_hidden;
+
+/* Avoid DWARF definition DIE on ifunc symbol so that GDB can handle
+   ifunc symbol properly.  */
+libc_ifunc (__memchr,
+	    (hwcap & PPC_FEATURE_HAS_VSX)
+            ? __memchr_power7
+            : __memchr_ppc);
+
+weak_alias (__memchr, memchr)
+libc_hidden_builtin_def (memchr)
+#else
+#include <string/memchr.c>
+#endif
-- 
1.8.3.1

